File: javascript/tests/client.test.ts
// Tests for StreamqClient using vitest's built-in fetch mocking.
// No real broker required.

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { StreamqClient } from "../src/client.js";
import {
  StreamqConflict,
  StreamqNotFound,
  StreamqAPIError,
} from "../src/errors.js";

function mockFetch(status: number, data: unknown) {
  return vi.spyOn(globalThis, "fetch").mockResolvedValueOnce({
    status,
    json: async () => data,
    ok: status >= 200 && status < 300,
  } as Response);
}

function brokerSuccess(data: unknown, status = 200) {
  return { status: "success", data };
}

function brokerError(message: string, status: number) {
  return mockFetch(status, { status: "error", message });
}

let client: StreamqClient;

beforeEach(() => {
  client = new StreamqClient("http://localhost:8080");
});

afterEach(() => {
  vi.restoreAllMocks();
});

describe("StreamqClient", () => {
  it("throws if baseUrl is empty", () => {
    expect(() => new StreamqClient("")).toThrow();
  });

  it("strips trailing slash from baseUrl", () => {
    const c = new StreamqClient("http://localhost:8080/");
    expect((c as any).baseUrl).toBe("http://localhost:8080");
  });
});

describe("createTopic", () => {
  it("succeeds on 201", async () => {
    mockFetch(201, brokerSuccess({ name: "payments" }));
    await expect(client.createTopic("payments")).resolves.toBeUndefined();
  });

  it("throws StreamqConflict on 409", async () => {
    brokerError("topic already exists", 409);
    await expect(client.createTopic("payments")).rejects.toThrow(
      StreamqConflict,
    );
  });

  it("sends correct method and path", async () => {
    const spy = mockFetch(201, brokerSuccess({ name: "payments" }));
    await client.createTopic("payments");
    expect(spy).toHaveBeenCalledWith(
      "http://localhost:8080/topics",
      expect.objectContaining({ method: "POST" }),
    );
  });
});

describe("deleteTopic", () => {
  it("succeeds on 200", async () => {
    mockFetch(200, brokerSuccess(null));
    await expect(client.deleteTopic("payments")).resolves.toBeUndefined();
  });

  it("throws StreamqNotFound on 404", async () => {
    brokerError("topic not found", 404);
    await expect(client.deleteTopic("missing")).rejects.toThrow(
      StreamqNotFound,
    );
  });
});

describe("listTopics", () => {
  it("returns parsed TopicInfo array", async () => {
    mockFetch(
      200,
      brokerSuccess([
        {
          name: "payments",
          message_count: 10,
          subscriber_count: 2,
          latest_offset: 9,
        },
        {
          name: "events",
          message_count: 5,
          subscriber_count: 0,
          latest_offset: 4,
        },
      ]),
    );

    const topics = await client.listTopics();
    expect(topics).toHaveLength(2);
    expect(topics[0]?.name).toBe("payments");
    expect(topics[0]?.messageCount).toBe(10);
    expect(topics[1]?.name).toBe("events");
  });

  it("returns empty array when no topics", async () => {
    mockFetch(200, brokerSuccess([]));
    const topics = await client.listTopics();
    expect(topics).toEqual([]);
  });
});

describe("getTopic", () => {
  it("returns parsed TopicInfo", async () => {
    mockFetch(
      200,
      brokerSuccess({
        name: "payments",
        message_count: 42,
        subscriber_count: 1,
        latest_offset: 41,
      }),
    );

    const info = await client.getTopic("payments");
    expect(info.name).toBe("payments");
    expect(info.latestOffset).toBe(41);
  });

  it("throws StreamqNotFound on 404", async () => {
    brokerError("topic not found", 404);
    await expect(client.getTopic("missing")).rejects.toThrow(StreamqNotFound);
  });
});

describe("publish", () => {
  it("returns PublishResult with offset", async () => {
    mockFetch(
      202,
      brokerSuccess({
        offset: 7,
        topic: "payments",
        timestamp: "2025-01-01T00:00:00Z",
      }),
    );

    const payload = new TextEncoder().encode('{"amount":100}');
    const result = await client.publish("payments", payload);

    expect(result.offset).toBe(7);
    expect(result.topic).toBe("payments");
  });

  it("base64-encodes payload in request body", async () => {
    const spy = mockFetch(
      202,
      brokerSuccess({
        offset: 0,
        topic: "payments",
        timestamp: "2025-01-01T00:00:00Z",
      }),
    );

    const payload = new TextEncoder().encode("hello");
    await client.publish("payments", payload);

    const call = spy.mock.calls[0];
    const body = JSON.parse(call?.[1]?.body as string) as { payload: string };
    // "hello" base64-encodes to "aGVsbG8="
    expect(body.payload).toBe("aGVsbG8=");
  });

  it("throws StreamqNotFound on 404", async () => {
    brokerError("topic not found", 404);
    await expect(
      client.publish("missing", new Uint8Array([1, 2, 3])),
    ).rejects.toThrow(StreamqNotFound);
  });
});

File: rust/tests/client_test.rs
// Integration tests for StreamqClient using wiremock.
// wiremock spins up a real HTTP server in-process, no real broker needed.
// Tests verify the full request/response cycle including JSON parsing,
// status code mapping, and error type dispatch.

use wiremock::{
    matchers::{body_json_string, method, path},
    Mock, MockServer, ResponseTemplate,
};

use streamq::{Error, StreamqClient};

fn success(data: serde_json::Value) -> ResponseTemplate {
    ResponseTemplate::new(200).set_body_json(serde_json::json!({
        "status": "success",
        "data": data
    }))
}

fn success_status(status: u16, data: serde_json::Value) -> ResponseTemplate {
    ResponseTemplate::new(status).set_body_json(serde_json::json!({
        "status": "success",
        "data": data
    }))
}

fn error_resp(status: u16, message: &str) -> ResponseTemplate {
    ResponseTemplate::new(status).set_body_json(serde_json::json!({
        "status": "error",
        "message": message
    }))
}

#[tokio::test]
async fn create_topic_success() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/topics"))
        .respond_with(success_status(
            201,
            serde_json::json!({ "name": "payments" }),
        ))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    assert!(client.create_topic("payments").await.is_ok());
}

#[tokio::test]
async fn create_topic_conflict() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/topics"))
        .respond_with(error_resp(409, "topic already exists"))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let err = client.create_topic("payments").await.unwrap_err();

    assert!(matches!(err, Error::Conflict(_)));
    if let Error::Conflict(msg) = err {
        assert!(msg.contains("already exists"));
    }
}

#[tokio::test]
async fn delete_topic_success() {
    let server = MockServer::start().await;

    Mock::given(method("DELETE"))
        .and(path("/topics/payments"))
        .respond_with(success(serde_json::Value::Null))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    assert!(client.delete_topic("payments").await.is_ok());
}

#[tokio::test]
async fn delete_topic_not_found() {
    let server = MockServer::start().await;

    Mock::given(method("DELETE"))
        .and(path("/topics/missing"))
        .respond_with(error_resp(404, "topic not found"))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let err = client.delete_topic("missing").await.unwrap_err();
    assert!(matches!(err, Error::NotFound(_)));
}

#[tokio::test]
async fn list_topics_returns_parsed_vec() {
    let server = MockServer::start().await;

    Mock::given(method("GET"))
        .and(path("/topics"))
        .respond_with(success(serde_json::json!([
            { "name": "payments", "message_count": 10, "subscriber_count": 2, "latest_offset": 9 },
            { "name": "events",   "message_count":  5, "subscriber_count": 0, "latest_offset": 4 },
        ])))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let topics = client.list_topics().await.unwrap();

    assert_eq!(topics.len(), 2);
    assert_eq!(topics[0].name, "payments");
    assert_eq!(topics[0].message_count, 10);
    assert_eq!(topics[1].name, "events");
}

#[tokio::test]
async fn list_topics_empty_returns_empty_vec() {
    let server = MockServer::start().await;

    Mock::given(method("GET"))
        .and(path("/topics"))
        .respond_with(success(serde_json::json!([])))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let topics = client.list_topics().await.unwrap();
    assert!(topics.is_empty());
}

#[tokio::test]
async fn get_topic_returns_info() {
    let server = MockServer::start().await;

    Mock::given(method("GET"))
        .and(path("/topics/payments"))
        .respond_with(success(serde_json::json!({
            "name": "payments",
            "message_count": 42,
            "subscriber_count": 1,
            "latest_offset": 41
        })))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let info = client.get_topic("payments").await.unwrap();

    assert_eq!(info.name, "payments");
    assert_eq!(info.latest_offset, 41);
}

#[tokio::test]
async fn get_topic_not_found() {
    let server = MockServer::start().await;

    Mock::given(method("GET"))
        .and(path("/topics/missing"))
        .respond_with(error_resp(404, "topic not found"))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    assert!(matches!(
        client.get_topic("missing").await.unwrap_err(),
        Error::NotFound(_)
    ));
}

#[tokio::test]
async fn publish_returns_offset() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/topics/payments/publish"))
        .respond_with(ResponseTemplate::new(202).set_body_json(serde_json::json!({
            "status": "success",
            "data": {
                "offset": 7,
                "topic": "payments",
                "timestamp": "2025-01-01T00:00:00Z"
            }
        })))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    let result = client
        .publish("payments", b"{\"amount\":100}")
        .await
        .unwrap();

    assert_eq!(result.offset, 7);
    assert_eq!(result.topic, "payments");
}

#[tokio::test]
async fn publish_base64_encodes_payload() {
    let server = MockServer::start().await;

    // Verify the request body contains a base64-encoded payload.
    // "hello" base64-encodes to "aGVsbG8="
    Mock::given(method("POST"))
        .and(path("/topics/payments/publish"))
        .and(body_json_string(r#"{"payload":"aGVsbG8="}"#))
        .respond_with(ResponseTemplate::new(202).set_body_json(serde_json::json!({
            "status": "success",
            "data": { "offset": 0, "topic": "payments", "timestamp": "2025-01-01T00:00:00Z" }
        })))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    client.publish("payments", b"hello").await.unwrap();
}

#[tokio::test]
async fn publish_not_found() {
    let server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/topics/missing/publish"))
        .respond_with(error_resp(404, "topic not found"))
        .mount(&server)
        .await;

    let client = StreamqClient::new(server.uri()).unwrap();
    assert!(matches!(
        client.publish("missing", b"data").await.unwrap_err(),
        Error::NotFound(_)
    ));
}

#[test]
fn error_display_not_found() {
    let e = Error::NotFound("topic not found".to_string());
    assert!(e.to_string().contains("not found"));
}

#[test]
fn error_display_conflict() {
    let e = Error::Conflict("already exists".to_string());
    assert!(e.to_string().contains("conflict"));
}

#[test]
fn error_display_api() {
    let e = Error::Api {
        status: 500,
        message: "internal error".to_string(),
    };
    assert!(e.to_string().contains("500"));
}

#[test]
fn error_display_max_reconnects() {
    let e = Error::MaxReconnects {
        attempts: 5,
        message: "refused".to_string(),
    };
    assert!(e.to_string().contains("5"));
}
